File: day-3/python/solution.py
import string


def get_priority_score(item):
    PRIORITIES_SCORE = dict(zip(string.ascii_letters, list(range(1, 53))))
    return PRIORITIES_SCORE[item]


def get_priority_sum(rucksacks):
    sum = 0
    for rucksack in rucksacks:
        midpt = int(len(rucksack) / 2)
        left = rucksack[0:midpt]
        right = rucksack[midpt:]
        misplaced_item = set(left).intersection(set(right)).pop()
        sum += get_priority_score(misplaced_item)
    return sum


def get_group_priority_sum(rucksacks, group_size):
    sum = 0
    for i in range(0, len(rucksacks), group_size):
        group_rucksacks = [set(rucksacks[i]) for i in range(i, i + group_size)]
        badge_set = group_rucksacks[0]
        for item_set in group_rucksacks[1:]:
            badge_set = badge_set.intersection(item_set)
        if len(badge_set) == 1:
            sum += get_priority_score(badge_set.pop())
    return sum


def display(sum):
    print("Priorities sum is {:,}.".format(sum))


with open('../inputs/input.txt', 'r') as f:
    rucksacks = f.read().splitlines()
    display(get_priority_sum(rucksacks))
    display(get_group_priority_sum(rucksacks, 3))

File: day-3/ruby/solution.rb
def sum_group(rucksacks)
  rucksacks.reduce(0) do |sum, rucksack|
    left = (rucksack[0...(rucksack.size / 2)]).chars
    right = (rucksack[(rucksack.size / 2)..]).chars
    item = (left & right).first
    sum + get_score(item)
  end
end

def sum_three_groups(rucksacks)
  rucksacks.each_slice(3).reduce(0) do |sum, slice|
    trio = slice.map(&:chars)
    sum + get_score((trio[0] & trio[1] & trio[2]).first)
  end
end

def get_score(item)
  item.ord - (item.ord <= 90 ? 38 : 96)
end

def display_total(sum)
  puts "The sum of common items was: #{sum}"
end

rucksacks = File.open('../inputs/input.txt').read.split("\n")
display_total(sum_group(rucksacks))
display_total(sum_three_groups(rucksacks))